src/python/pants/engine/internals/target_adaptor.py

```diff
@@ -10,7 +10,70 @@
 from typing_extensions import final
 
 from pants.build_graph.address import Address
+from pants.engine.collection import Collection
 from pants.engine.engine_aware import EngineAwareParameter
+from pants.util.frozendict import FrozenDict
+from pants.vcs.hunk import TextBlock
+
+
+@dataclass(frozen=True)
+class SourceBlock:
+    """Block of lines in a file.
+
+    Lines are 1 indexed, `start` is inclusive, `end` is exclusive.
+
+    SourceBlock is used to describe a set of source lines that are owned by a Target,
+    thus it can't be empty, i.e. `start` must be less than `end`.
+    """
+
+    start: int
+    end: int
+
+    def __init__(self, start: int, end: int):
+        object.__setattr__(self, "start", start)
+        object.__setattr__(self, "end", end)
+
+        self.__post_init__()
+
+    def __post_init__(self):
+        if self.start >= self.end:
+            raise ValueError(f"{self.start=} must be less than {self.end=}")
+
+    def __len__(self) -> int:
+        return self.end - self.start
+
+    def is_touched_by(self, o: TextBlock) -> bool:
+        """Check if the TextBlock touches the SourceBlock.
+
+        The function behaves similarly to range intersection check, but some edge cases are
+        different. See test cases for details.
+        """
+
+        if o.count == 0:
+            start = o.start + 1
+            end = start
+        else:
+            start = o.start
+            end = o.end
+
+        if self.end < start:
+            return False
+        if end < self.start:
+            return False
+        return True
+
+    @classmethod
+    def from_text_block(cls, text_block: TextBlock) -> SourceBlock:
+        """Convert (start, count) range to (start, end) range.
+
+        Useful for unified diff conversion, see
+        https://www.gnu.org/software/diffutils/manual/html_node/Detailed-Unified.html
+        """
+        return cls(start=text_block.start, end=text_block.start + text_block.count)
+
+
+class SourceBlocks(Collection[SourceBlock]):
+    pass
```

Comment:
From [this comment](https://github.com/pantsbuild/pants/pull/20767#discussion_r1606673735) I learned that we may want to have the order for this collection type not matter for caching purposes (as well as we don't expect to add/remove entries from this collection, right?) and can therefore preferably use the frozen ordered set. (as for `Owners` for instance)
https://github.com/pantsbuild/pants/blob/d81fd330043822a252e8cd08c0acf1822d69a418/src/python/pants/engine/internals/graph.py#L998-L999
```suggestion
class SourceBlocks(FrozenOrderedSet[SourceBlock]):
    pass
```

Reply:
Sure